src/ripple/app/book/impl/BookTip.cpp

```diff
@@ -46,31 +47,33 @@ BookTip::step ()
     for(;;)
     {
         // See if there's an entry at or worse than current quality.
-        auto const page (
-            view().getNextLedgerIndex (m_book, m_end));
+        auto const first_page (view().getNextLedgerIndex (m_book, m_end));
 
-        if (page.isZero())
+        if (first_page.isZero())
             return false;
 
         unsigned int di (0);
         SLE::pointer dir;
-        if (view().dirFirst (page, dir, di, m_index))
+
+        if (view().dirFirst (first_page, dir, di, m_index))
         {
             m_dir = dir->getIndex();
             m_entry = view().entryCache (ltOFFER, m_index);
+            m_quality = Quality (getQuality (first_page));
             m_valid = true;
```

Comment:
Aha! I didn't even notice that we set the quality on the first read - because this critical business logic change was intermixed with the refactoring...